src/rust-crypto/OutgoingRequestsManager.ts

```diff
@@ -0,0 +1,120 @@
+     * If there are additional calls to doProcessOutgoingRequests() while there is a current call running
+     * we need to remember in order to call process again (as there could be new requests).
+     */
+    private nextLoopDeferred?: IDeferred<void>;
+
+    public constructor(
+        private readonly logger: Logger,
+        private readonly olmMachine: OlmMachine,
+        public readonly outgoingRequestProcessor: OutgoingRequestProcessor,
+    ) {}
+
+    /**
+     * Shut down as soon as possible the current loop of outgoing requests processing.
+     */
+    public stop(): void {
+        this.stopped = true;
+    }
+
+    /**
+     * Process the OutgoingRequests from the OlmMachine.
+     *
+     * This should be called at the end of each sync, to process any OlmMachine OutgoingRequests created by the rust sdk.
+     * In some cases if OutgoingRequests need to be sent immediately, this can be called directly.
+     *
+     * Calls to doProcessOutgoingRequests() are processed synchronously, one after the other, in order.
+     * If doProcessOutgoingRequests() is called while another call is still being processed, it will be queued.
+     * Multiple calls to doProcessOutgoingRequests() when a call is already processing will be batched together.
+     */
+    public async doProcessOutgoingRequests(): Promise<void> {
+        if (this.isTaskRunning) {
+            // If the task is running, add the request to the queue and wait for completion.
+            // This ensures that the requests are processed only once at a time.
+            if (!this.nextLoopDeferred) {
+                this.nextLoopDeferred = defer();
+            }
+            return this.nextLoopDeferred.promise;
+        } else {
+            this.isTaskRunning = true;
+            try {
+                await this.processOutgoingRequests();
+            } finally {
+                this.isTaskRunning = false;
+            }
+
+            // If there was some request while this iteration was running, run a second time and resolve the linked promise.
+            if (this.nextLoopDeferred) {
+                if (this.stopped) {
+                    this.nextLoopDeferred.resolve();
+                    return;
+                }
+
+                // keep the current deferred requests to resolve them after the next iteration.
+                const deferred = this.nextLoopDeferred;
+                // reset the nextLoopDeferred so that any future requests are queued for another additional iteration.
+                this.nextLoopDeferred = undefined;
+
+                // Run again and resolve all the pending requests.
+                // Notice that we don't await on it, so that the current promise is resolved now.
+                // The requests that were deferred will be resolved after this new iteration.
+                this.doProcessOutgoingRequests().then(() => {
+                    deferred.resolve();
+                });
+            }
+        }
+    }
```

Comment:
Out of interest I tried rewriting this without the recursion.

IMHO it's clearer, but YMMV.

```suggestion
    public doProcessOutgoingRequests(): Promise<void> {
        // flag that we need at least one more iteration of the loop.
        if (!this.nextLoopDeferred) {
            this.nextLoopDeferred = defer();
        }
        
        // ... and wait for it to complete.
        let result = nextLoopDeferred.promise;
        
        // set the loop going if it is not already.
        if (!this.outgoingRequestLoopRunning) {
            this.outgoingRequestLoopInner().catch((e) => {
                // this should not happen; outgoingRequstLoopInner should return any errors via `nextLoopDeferred`.
                this.logger.error("Uncaught error in outgoing request loop", e);
            });
        }

        return result;
    } 
        
    private async outgoingRequestLoopInner(): Promise<void> {
        /* istanbul ignore if */
        if (this.outgoingRequestLoopRunning) {
            throw new Error("Cannot run two outgoing request loops");
        }        

        this.outgoingRequestLoopRunning = true;
        try {
            while (!this.stopped && this.nextLoopDeferred) {
                const deferred = this.nextLoopDeferred;
                
                // reset the nextLoopDeferred so that any future calls to doProcessOutgoingRequests are queued for another additional iteration.
                this.nextLoopDeferred = undefined;
                                
                // make the requests and feed the results back to the `nextLoopDeferred`
                await this.processOutgoingRequests().then(deferred.resolve, deferred.reject);
            }
        } finally {
             this.outgoingRequestLoopRunning = false;
         }
    }
```